keps/sig-scheduling/20190311-resource_bin_packing_priority_function.md

```diff
@@ -0,0 +1,175 @@
+---
+title: Resource Bin Packing Priority Function
+  - "@sudeshsh"
+owning-sig: sig-scheduling
+participating-sigs:
+  - sig-scheduling
+reviewers:
+  - "@k82cn"
+  - "@Huang-Wei"
+approvers:
+  - "@k82cn"
+creation-date: 2019-03-11
+last-updated: 2019-03-11
+status: provisional
+---
+
+# Resource Bin Packing Priority Function
+
+## Table of Contents
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+- [Design Details](#design-details)
+  - [Examples](#examples)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+
+## Summary
+
+Resource Bin Packing Priority Function allows users to use the best fit polices during scheduling. It allows users to apply [bin packing](https://en.wikipedia.org/wiki/Bin_packing_problem) on core resources like CPU, Memory as well as extended resources like accelerators.
+
+## Motivation
+
+While running Machine Learning workloads on kubernetes which use accelerator devices the default scheduler spreads pods across nodes resulting in fragmentation of extended resources. This fragmentation prevents scheduling pods with larger device resource requirements, and they remain in pending state.
+
+### Goals
+
+- Schedule Jobs using BestFit Policy using Resource Bin Packing Priority Function
+- Reduce Fragmentation of scarce resources on the Cluster
+
+### Non-Goals
+
+-
+
+## Proposal
+
+The plan is to modify  `requested_to_capacity_ratio` priority function to support an additional priority function `resource_bin_packing` along with `most_requested` and `least_requested`. Add another argument resources of type `map[v1.ResourceName]int64{}` .This would allow users who want to bin pack a resource to use the function by setting the argument resources which would require them to specify weights for bin packing. For example
+
+```yaml
+"priorities": [
+    ... {
+        "name": "ResourceBinPackingPriority",
+        "weight": 5,
+        "argument": {
+            "resourceBinPacking": {
+                "resources": [{
+                    "resource": "intel.com/foo",
```

Comment:
please highlight the behaviour on 1. resource no exist, 2. duplicated weight config.

Reply:
and is there any default resources, e.g. CPU, Memory?

Reply:
As `resources` is optional arguments, in case it's not passed it defaults to  ```go
resources := map[v1.ResourceName]int64{
  v1.ResourceMemory : 1,
  v1.ResourceCPU    : 1,
}
```

Reply:
that's ok to me, thanks :)